pg_shard.c

```diff
@@ -1004,6 +1407,215 @@ TargetEntryList(List *expressionList)
 
 
 /*
+ * CreateTemporaryTableLikeStmt returns a CreateStmt node depending on whether the
+ * aggregates in the localQuery can be pushed down to the workers or not.
+ */
+static CreateStmt *
+CreateTemporaryTableStmt(Query *distributedQuery, bool pushDownAggregates)
+{
+	CreateStmt *createStmt = NULL;
+
+	if (pushDownAggregates)
+	{
+		List *workerTargetList = distributedQuery->targetList;
+		List *rangeTableList = distributedQuery->rtable;
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) linitial(rangeTableList);
+		List *columnNameValueList = rangeTableEntry->eref->colnames;
+		List *columnNameList = ValueToStringList(columnNameValueList);
+		List *targetList = MasterTargetList(workerTargetList);
+		StringInfo tempTableName = CreateUniqueTableName();
+
+		createStmt = BuildCreateStatement(tempTableName, targetList, columnNameList);
+	}
+	else
+	{
+		Oid distributedTableId = ExtractFirstDistributedTableId(distributedQuery);
+
+		createStmt = CreateTemporaryTableLikeStmt(distributedTableId);
+	}
+
+	return createStmt;
+}
+
+
+/*
+ * ValueToStringList walks over the given list of string value types, converts
+ * value types to cstrings, and adds these cstrings into a new list.
+ */
+static List *
+ValueToStringList(List *valueList)
+{
+	List *stringList = NIL;
+	ListCell *valueCell = NULL;
+
+	foreach(valueCell, valueList)
+	{
+		Value *value = (Value *) lfirst(valueCell);
+		char *stringValue = strVal(value);
+
+		stringList = lappend(stringList, stringValue);
+	}
+
+	return stringList;
+}
+
+
+/*
+ * MasterTargetList uses the given worker target list's expressions, and creates
+ * a target target list for the master node. This master target list keeps the
+ * temporary table's columns on the master node.
+ */
+static List *
+MasterTargetList(List *workerTargetList)
```

Comment:
It's a bit confusing to mix Master/Local and Worker/Remote in the same file, but again we're stuck with whatever CitusDB calls it for this method.